DRAFT - SWE l1a algorithm

tech3371 opened the pull request:
# Change Summary

## Overview
This code is still in work. I wanted to open it in case anyone is working on similar task like mine and like to take a look at what I am doing.

## New Files
l1a_swe.py contains all of l1a algorithm.

## Updated Files
- pyproject.toml
   - It includes libraries that I needed locally to test but could be take away before final PR
- swe_packet_definition.xml
   - updated SWE packet definition to match their latest submission

## Testing
None

tech3371 committed:
fix merge

tech3371 committed:
fix merge

tech3371 committed:
committing swe l1a algorithm

tech3371 committed:
committing swe l1a algorithm

tech3371 committed:
update telemetry to match latest packet definition from TLM_SWE_20230904.xlsx

tech3371 committed:
update telemetry to match latest packet definition from TLM_SWE_20230904.xlsx

tech3371 committed:
removed converting to engineering unit in XTCE. TODO: check if we need to add it back in.

tech3371 committed:
removed converting to engineering unit in XTCE. TODO: check if we need to add it back in.

tech3371 committed:
some updates to l1a algorithm

tech3371 committed:
some updates to l1a algorithm

tech3371 committed:
undid some changes

tech3371 committed:
undid some changes

tech3371 committed:
undid codice changes.

tech3371 committed:
undid codice changes.

tech3371 committed:
code is little messy but I think it works. TODO: confirm it.

tech3371 committed:
code is little messy but I think it works. TODO: confirm it.

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+    spin_three_indexes = [[3,  7, 11, 15, 19,  23], [21, 17, 13, 9, 5, 1]]
+    spin_four_indexes = [[4,  8, 12, 16, 20,  24], [22, 18, 14, 10, 6, 2]]
+
+    match spin_number:
```
This is Python 3.10+, so I don't think we can use this quite yet if we want to keep supporting Python 3.9.

This could also be written as a dictionary lookup if you'd prefer:

```python
step_indices = {
    0: [[1,  5, 9, 13, 17,  21], [23, 19, 15, 11, 7, 3]],
    1: [[2,  6, 10, 14, 18,  22], [24, 20, 16, 12, 8, 4]],
    2: [[3,  7, 11, 15, 19,  23], [21, 17, 13, 9, 5, 1]],
    3: [[4,  8, 12, 16, 20,  24], [22, 18, 14, 10, 6, 2]]
}
```
Then access with `step_indices[0]`

tech3371 replied:
That's a great idea! I always forget about using dictionary.

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+from imap_processing.swe import decom_swe
+
+
+def uncompress(cem_count):
```
I'm not sure how often this is being used below, but you might want to consider adding a cache to this since it seems to be a direct lookup table calculation. Or just pre-computing the entire lookup table ahead of time:

```suggestion
@functools.cache
def uncompress(cem_count):
```

or since you know there is a maximum of 256 values that seems reasonable to store up front if you're going to be hitting it a lot.

`decompression_table = [uncompress(i) for i in range(256)]`

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+    index = cem_count // 16
+    multi = cem_count % 16
+
+    # uncompression formula
```
I'd suggest linking to explicit pages/locations in the algorithm documents in these formulas too. I think that will help us later going back to debug things and cross-reference check where values came from.

tech3371 replied:
will do

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+    elif 9 <= index <= 15:
+        return 2**index + 1024
+    else:
+        print("\nError! Switch for index failed in function N.")
+        return 0
```
```suggestion
        raise ValueError("Invalid base, must be between 0 and 15.")
```

tech3371 replied:
noted

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+
+        # Uncompress counts. Uncompressed data is a list of 1260
+        # where 1260 = 15 seconds x 7 CEMs x 12 energy steps
+        uncompressed_data = [uncompress(i) for i in byte_data]
```
Looks like you are calling this 1260 times on each packet, so I'd definitely suggest caching / creating the lookup table above.

tech3371 replied:
<img width="994" alt="Screen Shot 2023-09-06 at 8 53 54 AM" src="https://github.com/IMAP-Science-Operations-Center/imap_processing/assets/36522642/b3ffa2ec-8bdd-48df-bfe3-fcae04c676c0">
SWE lookup table looks like this.

I do agree that calculating ahead might be good but I will think about your suggestion offline to understand.

I plan to refactor code after discussion with SWE team this friday.

greglucas replied:
:+1: Sounds good. Looking at that table, is there a reason not to just hard-code those numbers into a list? I think you have some fancy 2^3 code currently, which is somewhat confusing when you're going back and forth between fixed table values. Unless the 2^3 is somewhere in the document as well.

tech3371 replied:
I was thinking that myself too - If I can take this table, hard code this base and step_size and look up using index value. I will ask them if this table will change.

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+        # packet's metadata.
+
+        # read raw data
+        binary_data = decom_data[total_packets].data["SCIENCE_DATA"].raw_value
+        # read raw data as binary array using bitstring
+        bit_array = bitstring.ConstBitStream(bin=binary_data)
+        # chunk binary into 1260 units each with 8-bits
+        byte_data = bit_array.readlist(["uint:8"] * 1260)
```
Alternatively if you wanted to go straight to a numpy array and avoid bitstrings since you are dealing with full bytes.
```suggestion
        binary_data = decom_data[total_packets].data["SCIENCE_DATA"].raw_value
        # read raw data into a uint8 numpy array
        byte_data = np.frombuffer(binary_data, dtype=np.uint8)
```

tech3371 replied:
good to know!

greglucas replied:
So, I decided to run some timing comparisons to see if this would be worth it actually...

```bash
python -m timeit 'import bitstring; bitstring.ConstBitStream(bytes(1260)).readlist(["uint:8"]*1260)'
100 loops, best of 5: 2.68 msec per loop
python -m timeit 'import numpy as np; np.frombuffer(bytes(1260), dtype=np.uint8)'
500000 loops, best of 5: 668 nsec per loop
```

That is a pretty massive improvement by using numpy directly. I'm kind of surprised bitstring is that slow...

tech3371 replied:
That's impressive!

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+        for every_second in range(15):
+            # go through first six steps of first half second and populate data in the right index
+            # of one_full_cycle array.
+            for every_step, index_number in enumerate(indexes[0]):
+                one_full_cycle[index_number-1, column_index] = reshaped_data[every_second, every_step]
```
Wow, they didn't make this easy on you! Nice work decoding all this.

I _think_ you might be able to get rid of the outer `every_second` loop, because `column_index` is incrementing with `every_second`?

```suggestion
            for every_step, index_number in enumerate(indexes[0]):
                one_full_cycle[index_number-1, ::2] = reshaped_data[:, every_step]
                one_full_cycle[index_number-1, 1::2] = reshaped_data[:, 6 + every_step]
```

tech3371 replied:
I will try that. I was trying to think different ways and it was mind twisting. I was happy when this thing worked but was hoping for suggestion like this. I am going to ask them how they did in the past or what we have is good or not.

greglucas replied:
If you're looking for more things to think about, I believe `take()` or `take_along_axis()` might be able to do the fancy numpy indexing for you to remove the Python loops altogether based on your index lookup tables.

https://numpy.org/doc/stable/reference/generated/numpy.take.html

I think it is very good to start with the basic for-loops here to make sure you have things working as you expect and then optimize one thing at a time, so I think you've got a good approach going right now.

tech3371 replied:
cool. I will take a look at that.

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+                list(one_full_cycle), dims=["energy_step", "cem_counts", "angle"],
+            )
+            dataset = xr.Dataset({"data": xarray_data})
+            dataset.to_netcdf(f"swe_l1a_{total_packets}.nc")
```
:( I wish we could use netcdf too! I don't think we should keep this as a dependency, unfortunately. One thought would be to add it to testing dependencies.

tech3371 replied:
This was mainly for testing. This line will be replaced by xarray_to_cdf call. I plan to remove it before opening the PR.

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+
+    total_packets = 0
+    one_full_cycle = np.zeros((24, 30, 7))
+    while total_packets < len(decom_data):
```
```suggestion
    for packet in decom_data:
```

tech3371 replied:
I had that originally but I don't remember now why I changed it to while statement. :) I went through many iteration of implementation.

greglucas commented on imap_processing/swe/l1a_swe.py:
```diff
@@ -0,0 +1,183 @@
+        total_packets += 1
+
+
+swe_l1a()
```
```suggestion
if __name__ == "__main__":
    swe_l1a()
```

greglucas reviewed (commented):
I like all the comments throughout! Very helpful to follow what is going on.

I only have pretty minor comments/suggestions for optimizations, so nothing critical at all. I appreciate the TODOs in the code as well knowing that we'll need to come back to those areas.

sdhoyt committed:
Issue management documentation (#85)

* first pass at reqs and issue tracking docs

* added new doc page to style guide index

* added more issue details

* fixed typos

* fixed code formatting

* fixed code formatting

* updated sub-issue description

* Update docs/source/development/style-guide/requirement-and-issue-tracking.rst

Co-authored-by: Maxine Hartnett <117409426+maxinelasp@users.noreply.github.com>

* Update docs/source/development/style-guide/requirement-and-issue-tracking.rst

Co-authored-by: Maxine Hartnett <117409426+maxinelasp@users.noreply.github.com>

* Update docs/source/development/style-guide/requirement-and-issue-tracking.rst

Co-authored-by: Maxine Hartnett <117409426+maxinelasp@users.noreply.github.com>

* added project-management section

* fixed formatting issues

* another formatting fix attempt

* trying * for bullet points again

* specified SDC requirements in project management page

* removed req page from style guide

* comment updates

* updated git issue lifecycle image

---------

Co-authored-by: Maxine Hartnett <117409426+maxinelasp@users.noreply.github.com>

sdhoyt committed:
SDC Test Plan docs (#88)

* first pass at reqs and issue tracking docs

* added new doc page to style guide index

* added more issue details

* fixed typos

* fixed code formatting

* fixed code formatting

* updated sub-issue description

* Update docs/source/development/style-guide/requirement-and-issue-tracking.rst

Co-authored-by: Maxine Hartnett <117409426+maxinelasp@users.noreply.github.com>

* Update docs/source/development/style-guide/requirement-and-issue-tracking.rst

Co-authored-by: Maxine Hartnett <117409426+maxinelasp@users.noreply.github.com>

* Update docs/source/development/style-guide/requirement-and-issue-tracking.rst

Co-authored-by: Maxine Hartnett <117409426+maxinelasp@users.noreply.github.com>

* added project-management section

* fixed formatting issues

* another formatting fix attempt

* trying * for bullet points again

* specified SDC requirements in project management page

* removed req page from style guide

* added sdc sit plan doc

* updated wording about SIT procedure page

* added finalize test report step

* backing out requirements docs for PR

* added link to SOC test plan

* fix link

* another SIT Test Plan link fix attempt

* final test plan link fix

* minor wording changes

* removed images from issue docs for PR

* comment updates

* linking headers to table

* table fix

* another table linking fix attempt

* spacing fix

* ruff fixes

---------

Co-authored-by: Maxine Hartnett <117409426+maxinelasp@users.noreply.github.com>

bryan-harter committed:
Init commit of IDEX packet parser (#79)

* Init commit of IDEX packet parser

* Committing test data

* Adding xarray as a dependency

* Responding to code review

* Fixing the dependencies

* updating poetry.lock to sync with main branch

* Getting rid of old import statement

* Updating poetry lock file

* Fixing the lock file, for some reason bitstring is
still causing issues

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Updating to latest version of space packet parser

* Changing packet parser to use individual Dust
Impact classes and output an xarray object

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Changing the requirements in poetry for xarray

* Responding to code review comments, adding a unit
test to compare actual data

* [pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

* Updating poetry.lock

* Logging an error if a new science type is received

* Determining High and Low sample now
(rather than placeholder)

* Add more comments and make the variables clearer

* Fixing some of the variable names, adding comments

* Bug fix, wasn't returning right thing in new func

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>

GFMoraga committed:
Adding CoDICE Event data + Species counts packet definitions for L0 (#106)

"made `telemetry_generator.py` more user friendly, updated with BinaryParameterType, shorted codice/L0/ files"

tech3371 committed:
Merge branch 'dev' of github.com:tech3371/imap_processing into dev

tech3371 commented:
I am closing this draft PR because this code is going to change a lot after my discussion with SWE team this afternoon.
